Framework/Algorithms/src/FFTPreProcessing.cpp

```
@@ -0,0 +1,228 @@
+  const int padding = getProperty("Padding");
+  fptr apodizationFunction = getApodizationFunction(method);
+  // Do the specified spectra only
+  int specLength = static_cast<int>(spectra.size());
+  std::vector<double> norm(specLength, 0.0);
+  PARALLEL_FOR_IF(Kernel::threadSafe(*inputWS, *outputWS))
+  for (int i = 0; i < specLength; ++i) {
+    PARALLEL_START_INTERUPT_REGION
+    const auto specNum = static_cast<size_t>(spectra[i]);
+
+    if (spectra[i] > static_cast<int>(numSpectra)) {
+      g_log.error("The spectral index " + std::to_string(spectra[i]) +
+                  " is greater than the number of spectra!");
+      throw std::invalid_argument("The spectral index " +
+                                  std::to_string(spectra[i]) +
+                                  " is greater than the number of spectra!");
+    }
+    // Create output ws
+    outputWS->setHistogram(specNum,
+                           applyApodizationFunction(
+                               addPadding(inputWS->histogram(specNum), padding),
+                               decayConstant, apodizationFunction));
+    prog.report();
+    PARALLEL_END_INTERUPT_REGION
+  }
+  PARALLEL_CHECK_INTERUPT_REGION
+  setProperty("OutputWorkspace", outputWS);
+}
+
+typedef double (*fptr)(const double time, const double decayConstant);
+/**
+* Gets a pointer to the relevant
+* apodization function
+* @param method :: [input] The name of the chosen function
+* @returns :: pointer to the function
+*/
+fptr FFTPreProcessing::getApodizationFunction(const std::string method) {
+  if (method == "None") {
+    return none;
+  } else if (method == "Bartlett") {
+    return bartlett;
+  } else if (method == "Lorentz") {
+    return lorentz;
+  } else if (method == "Connes") {
+    return connes;
+  } else if (method == "Cosine") {
+    return cosine;
+  } else if (method == "Gaussian") {
+    return gaussian;
+  } else if (method == "Welch") {
+    return welch;
+  }
+  return none;
```

An exception is preferable here instead of silently returning none.